exonum/src/runtime/dispatcher/mod.rs

```diff
@@ -373,57 +421,61 @@ impl Dispatcher {
         &self,
         instance_id: InstanceId,
     ) -> Option<(u32, &dyn Runtime)> {
-        let ServiceInfo { runtime_id, .. } = self.service_infos.get(&instance_id)?;
+        let runtime_id = self
+            .service_infos
+            .get_runtime_for_active_instance(instance_id)?;
         let runtime = self.runtimes[&runtime_id].as_ref();
-        Some((*runtime_id, runtime))
+        Some((runtime_id, runtime))
     }
 
     /// Returns the service matching the specified query.
     pub(crate) fn get_service<'s>(
         &'s self,
-        fork: &Fork,
         id: impl Into<InstanceQuery<'s>>,
     ) -> Option<InstanceDescriptor<'s>> {
-        match id.into() {
-            InstanceQuery::Id(id) => {
-                let name = self.service_infos.get(&id)?.name.as_str();
-                Some(InstanceDescriptor { id, name })
-            }
-
-            InstanceQuery::Name(name) => {
-                // TODO: This may be slow.
-                let id = Schema::new(fork).instances().get(name)?.spec.id;
-                Some(InstanceDescriptor { id, name })
-            }
-        }
+        self.service_infos
+            .get_instance(id)
+            .and_then(|(descriptor, status)| {
+                if status.is_active() {
```

Comment:
This pattern has several occurrences within code. Maybe it's possible to move it into some method, like `descriptor.from_status(status)`?